docs/rfcs/IntelGpuDialect.md

```diff
@@ -0,0 +1,122 @@
+
+```
+
+func.func @main() {
+    %c8 = arith.constant 8 : index
+    %c1 = arith.constant 1 : index
+    %cst = arith.constant 2.200000e+00 : f32
+    %cst_0 = arith.constant 1.100000e+00 : f32
+    %cst_1 = arith.constant 0.000000e+00 : f32
+    %0 = "intel_gpu.create_stream"() : () -> !intel_gpu.OpaqueType
+    .
+    .
+    .
+    .
+    %6 = gpu.launch_func async [%5] @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%memref : memref<8xf32>, %memref_2 : memref<8xf32>, %memref_4 : memref<8xf32>)
+    %7 = memref.cast %memref_4 : memref<8xf32> to memref<*xf32>
+    gpu.wait [%6]
+    call @printMemrefF32(%7) : (memref<*xf32>) -> ()
+    "intel_gpu.destroy_stream"(%0) : (!intel_gpu.OpaqueType) -> ()
+    return
+  }
+
+  ```
+
+After lowering to LLVM the IR will be transformed into:
+
+```
+
+func.func @main() {
+    %c8 = arith.constant 8 : index
+    %0 = builtin.unrealized_conversion_cast %c8 : index to i64
+    %c1 = arith.constant 1 : index
+    %1 = builtin.unrealized_conversion_cast %c1 : index to i64
+    %cst = arith.constant 2.200000e+00 : f32
+    %cst_0 = arith.constant 1.100000e+00 : f32
+    %cst_1 = arith.constant 0.000000e+00 : f32
+    %3 = llvm.call @iGpuCreateStream() : (!llvm.ptr<i8>) -> !llvm.ptr<i8>
+    %4 = llvm.mlir.constant(8 : index) : i64
+    %5 = llvm.mlir.constant(1 : index) : i64
+    %6 = llvm.mlir.null : !llvm.ptr<f32>
+    %7 = llvm.getelementptr %6[%4] : (!llvm.ptr<f32>, i64) -> !llvm.ptr<f32>
+    %8 = llvm.ptrtoint %7 : !llvm.ptr<f32> to i64
+    %9 = llvm.mlir.constant(64 : i64) : i64
+    %10 = llvm.call @iGpuMemAlloc(%8, %9, %3) : (i64, i64, !llvm.ptr<i8>) -> !llvm.ptr<i8>
+    .
+    .
+    .
+    %62 = llvm.call @iGpuModuleLoad(%3, %60, %61) : (!llvm.ptr<i8>, !llvm.ptr<i8>, i64) -> !llvm.ptr<i8>
+    .
+    %66 = llvm.call @iGpuKernelGet(%3, %62, %65) : (!llvm.ptr<i8>, !llvm.ptr<i8>, !llvm.ptr<i8>) -> !llvm.ptr<i8>
+    .
+    llvm.call @iGpuLaunchKernel(%3, %66, %0, %1, %1, %1, %1, %1, %67, %108, %109) : (!llvm.ptr<i8>, !llvm.ptr<i8>, i64, i64, i64, i64, i64, i64, i32, !llvm.ptr<struct<(ptr<i8>, i64)>>, !llvm.ptr<ptr<i8>>) -> ()
+    .
+    .
+    llvm.call @iGpuDestroyStream(%3) : (!llvm.ptr<i8>) -> ()
+    return
+  }
+```
+
+## Features to be added in subsequent PR's (either in the dialect or via a pass operating on this dialect):
+1. More ops can be added to dialect going forward based on the use cases and requirements.
+2. Caching of the module creation needs to be added via a pass working on this dialect.
+3. Currenlty, the stream creation and destruction happens for every function. A pass to manage stream will be added where the stream is created only once at the beginning of the program and used throughout.
+
+## Alternative
+
+The alternative approch here is to add this operation in the upstream dialect and extend the upstream GPU dialect. But, we feel that having our own dialect will give us more command and control over how we want to design it.
```

Comment:
I would expect the upstream GPU would evolve to expose stream support to GPU dialect user.

Reply:
> > The create/destroy and using stream for individual kernel launch is fine level control than the region idea. A region using one stream/context can be supported by this programming model. There are other use cases like the user like to use different stream to launch different kernel.
> 
> I don't understand. You can have multiple regions as you can have multiple create-streams. Where is the difference in granularity? Where would the region reside?

Please note that this PRC needs to be changed to "User decide which stream to use for each kernel, not the compiler".  This give user fine-level control.  The region may also work but region usually contain multiple kernels dispatching to one same stream.